jenkins/snapshot-elasticsearch.sh

```diff
@@ -0,0 +1,137 @@
+        "       delete    deletes the named snapshots."
+        ''
+        "Example: $av0 dev list"
+        ''
+        "BTW, you ran: $*"
+        ''
+    )
+    local ok=no x
+    for x in ${environments[*]}; do test $x = "$environment" && ok=yes; done
+    if test $ok = no
+    then
+        echo 1>&2 "${av0}: Unrecognized environment: '$environment'"
+    fi
+    ok=no
+    for x in ${verbs[*]}; do test $x = "$verb" && ok=yes; done
+    if test $ok = no
+    then
+        echo 1>&2 "${av0}: Unrecognized verb: '$verb'"
+    fi
+    for x in "${tools[@]}"; do which $x >/dev/null || ok=no; done
+    if test $ok = no
+    then
+        echo 1>&2 "${av0}: You need these tools on PATH: ${tools[*]}"
+    fi
+    if test $ok = no
+    then
+        for line in "${usage[@]}"; do echo "$line" 1>&2; done
+        exit 1
+    fi
+}
+
+# Report that $av0 is running the rest of arguments as a command line.
+# Then run that command line and pipe it through 'jq .'.
+#
+function echoRunJq () {
+    local -r av0="$1" ; shift
+    local -a command
+    local arg result
+    for arg in "$@"; do command+=("'$arg'"); done
+    echo $av0: Running "${command[@]}"
+    result=$("$@")
+    echo "$result" | jq .
+}
+
+# Configure snapshots to $bucket in for the $es cluster.
+#
+function snapshot_configure () {
+    local -r av0="$1" es=$2 bucket=$3
+    local -r data='
+    {
+        "type": "gcs",
+        "settings": {
+            "bucket": "'$bucket'",
+            "service_account": "_default_",
+            "compress": true,
+            "application_name": "clio-elasticsearch"
+        }
+    }'
+    echoRunJq "$av0" curl -s -X PUT --data "$data" $es/_snapshot/$bucket
+}
+
+# List the snapshots in $environment for $es cluster.
+#
+function snapshot_list () {
+    local -r av0="$1" es=$2 bucket=$3
+    echoRunJq "$av0" curl -s $es/_snapshot/$bucket/_all
+}
+
+# Take a snapshot in $environment for $es cluster.
+#
+function snapshot_take () {
+    local -r av0="$1" es=$2 bucket=$3
+    local -r jq='keys|.[]|select(startswith(".")|not)'
+    local -r snapshot=clio-$(date -u +%Y-%m-%d-%H-%M-%S)
+    local -r url=$es/_snapshot/$bucket/$snapshot?wait_for_completion=true
+    local -a -r curl=(curl -s $es/_aliases)
+    local -r indexes=$("${curl[@]}" | jq --raw-output $jq)
+    if test -z "$indexes"
+    then
+        echo 1>&2 $av0: No document indexes in $es
+        echoRunJq "$av0" "${curl[@]}"
+        exit 2
+    fi
+    local index indices comma
+    for index in $indexes; do indices=$indices$comma$index; comma=,; done
+    local -r data='
+    {
+        "indices": "'$indices'",
+        "include_global_state": false
+    }'
+    echoRunJq "$av0" curl -s -X PUT --data "$data" $url
+}
+
+# Delete the snapshots named in $5 ... and so on.
+#
+function snapshot_delete () {
+    local -r av0="$1" es=$2 bucket=$3 ; shift 3
+    for snap in "$@"
+    do
+        echoRunJq "$av0" curl -s -X DELETE $es/_snapshot/$bucket/$snap
+    done
+}
+
+function main () {
```

Comment:
Not sure if the name `snapshot` is best here as this does a few things. When we have a restore operation, would it go here as well? Or another script?

Reply:
Suggest a better name.